src/hotspot/share/gc/parallel/psClosure.inline.hpp

```diff
@@ -77,63 +77,52 @@ class PSRootsClosure: public OopClosure {
 typedef PSRootsClosure</*promote_immediately=*/false> PSScavengeRootsClosure;
 typedef PSRootsClosure</*promote_immediately=*/true> PSPromoteRootsClosure;
 
-// Scavenges a single oop in a ClassLoaderData.
-class PSScavengeFromCLDClosure: public OopClosure {
-private:
+// Scavenges the oop in a ClassLoaderData.
+class PSScavengeCLDClosure: public CLDClosure {
   PSPromotionManager* _pm;
-  // Used to redirty a scanned cld if it has oops
-  // pointing to the young generation after being scanned.
-  ClassLoaderData*    _scanned_cld;
 public:
-  PSScavengeFromCLDClosure(PSPromotionManager* pm) : _pm(pm), _scanned_cld(nullptr) { }
-  void do_oop(narrowOop* p) { ShouldNotReachHere(); }
-  void do_oop(oop* p)       {
-    ParallelScavengeHeap* psh = ParallelScavengeHeap::heap();
-    assert(!psh->is_in_reserved(p), "GC barrier needed");
-    if (PSScavenge::should_scavenge(p)) {
-      assert(PSScavenge::should_scavenge(p, true), "revisiting object?");
+  PSScavengeCLDClosure(PSPromotionManager* pm) : _pm(pm) { }
 
-      oop o = RawAccess<IS_NOT_NULL>::oop_load(p);
-      oop new_obj = _pm->copy_to_survivor_space</*promote_immediately=*/false>(o);
-      RawAccess<IS_NOT_NULL>::oop_store(p, new_obj);
-
-      if (PSScavenge::is_obj_in_young(new_obj)) {
-        do_cld_barrier();
-      }
+  void do_cld(ClassLoaderData* cld) {
+    // If the cld has not been dirtied we know that there are
+    // no references into the young gen, so we can skip it.
+    if (!cld->has_modified_oops()) {
+      return;
     }
-  }
 
-  void set_scanned_cld(ClassLoaderData* cld) {
-    assert(_scanned_cld == nullptr || cld == nullptr, "Should always only handling one cld at a time");
-    _scanned_cld = cld;
-  }
+    // Scavenges a single oop in a ClassLoaderData.
+    class CLDOopClosure : public OopClosure {
```

Comment:
This makes the do_cld function much harder to read, IMHO. So, three alternatives:
1) Keep the class outside PSScavengeCLDClosure, just like it was before this patch.
2) Have the class inside PSScavengeCLDClosure, but not in a function
3) Have the class inside a function inside PSScavengeCLDClosure.

I prefer (1) because it allows you to look at PSScavengeCLDClosure without having to see the details about the second class. I'm OK with (2) and like that it encapsulates the other class. I don't think we should do (3) because it significantly hurts the readability of do_cld.

Reply:
I have moved the oop-closure class def outside.